src/libopensc/pkcs15-skeid.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * PKCS15 emulation layer for Slovak eID card
+ *
+ * Copyright (C) 2022 Juraj Šarinay <juraj@sarinay.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ * based on the PKCS15 emulation layer for EstEID card by Martin Paljak
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "common/compat_strlcpy.h"
+
+#include "internal.h"
+#include "log.h"
+#include "pkcs15.h"
+
+static const struct sc_aid skeid_aid_qes = {{0xE8, 0x28, 0xBD, 0x08, 0x0F, 0xA0, 0x00, 0x00, 0x08, 0x51, 0x00, 0x00, 0x11}, 13};
+
+static int sc_pkcs15emu_skeid_init(sc_pkcs15_card_t * p15card)
+{
+	int r;
+	int i;
+	int sn_len;
+	char *buf;
+
+	set_string(&p15card->tokeninfo->label, "eID karta");
+	set_string(&p15card->tokeninfo->manufacturer_id, "Ministerstvo vnutra SR");
+
+	sn_len = p15card->card->serialnr.len;
+	if (sn_len > 0) {
+		buf = malloc(2 * sn_len + 1);
+		if (!buf) return SC_ERROR_OUT_OF_MEMORY;
+		sc_bin_to_hex(p15card->card->serialnr.value, sn_len, buf,
+			2 * sn_len + 1, 0);
+		p15card->tokeninfo->serial_number = buf;
+	}
+
+	p15card->tokeninfo->flags = SC_PKCS15_TOKEN_PRN_GENERATION | SC_PKCS15_TOKEN_READONLY;
+
+	for (i = 0; i < 3; i++) {
+		const char *skeid_cert_names[3] = {
+			"Kvalifikovany certifikat",
+			"Podpisovy certifikat",
+			"Sifrovaci certifikat"
+		};
+
+		const char *skeid_cert_paths[3] = {
+			"3f0001030201",
+			"3f0001030202",
+			"3f0001030203"
+		};
+
+		struct sc_pkcs15_cert_info cert_info;
+		struct sc_pkcs15_object cert_obj;
+
+		memset(&cert_info, 0, sizeof(cert_info));
+		memset(&cert_obj, 0, sizeof(cert_obj));
+
+		cert_info.id.value[0] = i + 1;
+		cert_info.id.len = 1;
+
+		sc_format_path(skeid_cert_paths[i], &cert_info.path);
+		strlcpy(cert_obj.label, skeid_cert_names[i], sizeof(cert_obj.label));
+
+		r = sc_pkcs15emu_add_x509_cert(p15card, &cert_obj, &cert_info);
+
+		LOG_TEST_RET(p15card->card->ctx, r, "Error adding certificate.");
+	}
+
+	/* add pins */
+	for (i = 0; i < 2; i++) {
+		const char *skeid_pin_names[2] = {
+			"BOK",
```

Comment:
Are the terms widely used? Or are they from specification? See my comment above - I know from experience that as these values are presented to users, they better be "average joe readable" (even if not precise)